packages/@sanity/form-builder/src/inputs/ReferenceInput/ReferenceInput.tsx

```diff
@@ -1,336 +1,311 @@
+import {Details} from '../../components/Details'
+import {IntentButton} from '../../transitional/IntentButton'
+import {EMPTY_OBJECT} from '../../utils/empty'
+import {usePreviewSnapshot} from './usePreviewSnapshot'
 
-type SearchHit = {
-  _id: string
-  _type: string
+type SearchState = {
+  hits: SearchHit[]
+  isLoading: boolean
 }
+
+const INITIAL_SEARCH_STATE: SearchState = {
+  hits: [],
+  isLoading: false,
+}
+
 type PreviewSnapshot = {
+  _id: string
   _type: string
   title: string
   description: string
 }
 
-type SearchError = {
-  message: string
-  details?: {
-    type: string
-    description: string
-  }
-}
+type SearchFunction = (query: string) => Observable<SearchHit[]>
+
 export type Props = {
   value?: Reference
   compareValue?: Reference
   type: ReferenceSchemaType
   markers: Marker[]
   focusPath: Path
   readOnly?: boolean
-  onSearch: (
-    query: string,
-    type: ReferenceSchemaType,
-    options: ReferenceFilterSearchOptions
-  ) => Observable<SearchHit[]>
-  onFocus: (path: Path) => void
-  getPreviewSnapshot: (reference, type) => Observable<PreviewSnapshot>
+  onSearch: SearchFunction
+  onFocus?: (path: Path) => void
+  onBlur?: () => void
+  getPreviewSnapshot: (reference: Reference) => Observable<PreviewSnapshot | null>
   onChange: (event: PatchEvent) => void
   level: number
-  presence: any
-
-  // From withDocument
-  document: SanityDocument
+  presence: FormFieldPresence[]
+}
 
-  // From withValuePath
-  getValuePath: () => Path
+function getMemberTypeFor(typeName: string, ownerType: ReferenceSchemaType) {
+  return ownerType.to.find((ofType) => ofType.type?.name === typeName)
 }
 
-type State = {
-  isFetching: boolean
-  hits: Array<SearchHit>
-  isMissing: boolean
-  previewSnapshot: PreviewSnapshot | null
-  refCache: {
-    [key: string]: SearchHit
-  }
+type SearchHit = {
+  _id: string
+  _type: string
 }
-const getInitialState = (): State => {
-  return {
-    isFetching: false,
-    hits: [],
-    previewSnapshot: null,
-    isMissing: false,
-    refCache: {},
-  }
+
+const NO_FILTER = () => true
+
+function nonNullable<T>(v: T): v is NonNullable<T> {
+  return v !== null
 }
 
-export default withValuePath(
-  withDocument(
-    class ReferenceInput extends React.Component<Props, State> {
-      _lastQuery = ''
-      _input: SearchableSelect
-      state = getInitialState()
-      subscriptions = subscriptionManager('search', 'previewSnapshot')
-      _inputId = uniqueId('ReferenceInput')
-      componentWillUnmount() {
-        this.subscriptions.unsubscribeAll()
-      }
-      componentDidMount() {
-        this.getPreviewSnapshot(this.props.value)
-      }
+export const ReferenceInput = forwardRef(function ReferenceInput(
+  props: Props,
+  forwardedRef: ForwardedRef<HTMLInputElement>
+) {
+  const {
+    type,
+    value,
+    level,
+    markers,
+    readOnly,
+    onSearch,
+    onChange,
+    presence,
+    compareValue,
+    focusPath = [],
```

Comment:
This will create a new empty array on every render, when no `focusPath` is provided

Reply:
Good catch, thanks!